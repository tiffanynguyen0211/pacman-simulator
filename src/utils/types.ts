export type Direction = 'west' | 'south' | 'east' | 'north';

export type PacManPosition = {
    col: number;
    row: number;
    direction: Direction;
};

export type Grid = {
    rows: number;
    cols: number;
    grid: {
        row: number;
        col: number;
    }[];
    width: number;
    height: number;
    unit: number;
    spacing: number;
};
