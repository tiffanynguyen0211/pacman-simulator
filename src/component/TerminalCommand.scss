.terminal-container {
  width: 100%;
  margin-bottom: 20px;
}
  
.terminal-result {
  overflow: auto;
  width: 100%;
  height: 60%;
  background-color: rgb(31, 30, 30);
  color: white;
  text-align: left;

p {
  margin: 10px;
}
}
  
.terminal-input{
  width: 100%;
  padding: 10px;
  border-top: 2px solid grey;
  background-color: rgb(31, 30, 30);
  color: white;
}

.terminal-container {
  display: flex;
  width: 100vw;
  height: 50vh;
  flex-direction: column;
  align-items: center;
  justify-content: center;
}
  