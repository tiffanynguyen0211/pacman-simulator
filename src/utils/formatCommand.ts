import { Direction, PacManPosition } from './types';

export const placeCommandRegex = /^place\s[0-9],[0-9],(north|west|east|south)$/;

export const formattedPlaceCommand = (
    text: string,
): { row: number; col: number; direction: Direction } => {
    const newPosition = text.substr(6);
    const row = Number(newPosition.split(',')[0]);
    const col = Number(newPosition.split(',')[1]);
    const direction = newPosition.split(',')[2] as Direction;

    return { row, col, direction };
};

export const printReport = ({
    row,
    col,
    direction,
}: PacManPosition): string => {
    return `${col},${row},${direction}`;
};

export const turnLeftDirection = (direction: Direction): Direction => {
    switch (direction) {
        case 'east': {
            return 'north';
        }
        case 'north': {
            return 'west';
        }
        case 'west': {
            return 'south';
        }
        default: {
            return 'east';
        }
    }
};

export const turnRightDirection = (direction: Direction): Direction => {
    switch (direction) {
        case 'east': {
            return 'south';
        }
        case 'south': {
            return 'west';
        }
        case 'west': {
            return 'north';
        }
        default: {
            return 'east';
        }
    }
};
