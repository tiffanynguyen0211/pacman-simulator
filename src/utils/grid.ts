import { Grid } from './types';

const unit = 50;
const spacing = 5;
const rows = 5;
const cols = 5;

const isWithinGrid = (rowIndex: number, colIndex: number): boolean => {
    return (
        colIndex >= 0 &&
        colIndex <= cols - 1 &&
        rowIndex >= 0 &&
        rowIndex <= rows - 1
    );
};

const gridHeight = cols * (unit + 2 * spacing);
const gridWidth = rows * (unit + 2 * spacing);

const initGrid = (): Grid => {
    const grid = [];
    for (let row = 0; row < rows; row += 1) {
        for (let col = 0; col < cols; col += 1) {
            grid.push({
                row,
                col,
            });
        }
    }

    return {
        rows,
        cols,
        grid: grid.reverse(),
        width: gridWidth,
        height: gridHeight,
        unit,
        spacing,
    };
};

export { initGrid, isWithinGrid };
