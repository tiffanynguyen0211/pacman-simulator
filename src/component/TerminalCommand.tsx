import './TerminalCommand.scss';

import { KeyboardEvent, useRef } from 'react';

import {
    formattedPlaceCommand,
    placeCommandRegex,
    printReport,
    turnLeftDirection,
    turnRightDirection,
} from '../utils/formatCommand';
import { isWithinGrid } from '../utils/grid';
import { PacManPosition } from '../utils/types';

export const TerminalCommand: React.FC<{
    pacMan: PacManPosition;
    setPacMan: (newPacmanPosition: PacManPosition) => void;
}> = ({ pacMan, setPacMan }) => {
    const inputRef = useRef<HTMLInputElement>(null);
    const terminalResultRef = useRef<HTMLDivElement>(null);

    function logCommandHistory(text: string) {
        const terminal = terminalResultRef.current;
        if (terminal) {
            terminal.innerHTML += `<p>${text.toUpperCase()}</p>`;
            // scroll to latest terminal log
            terminal.scrollTop = terminal.scrollHeight;
        }
    }
    function clearInput() {
        if (inputRef && inputRef.current) {
            inputRef.current.value = '';
        }
    }

    function checkCommand(text: string) {
        clearInput();
        switch (true) {
            case text === 'report': {
                const result = printReport(pacMan);
                logCommandHistory(`<p>Output: ${result.toUpperCase()}</p>`);
                break;
            }
            case text === 'left': {
                const leftDirection = turnLeftDirection(pacMan.direction);
                setPacMan({
                    ...pacMan,
                    direction: leftDirection,
                });
                break;
            }
            case text === 'right': {
                const rightDirection = turnRightDirection(pacMan.direction);
                setPacMan({
                    ...pacMan,
                    direction: rightDirection,
                });
                break;
            }
            case placeCommandRegex.test(text): {
                const {
                    row,
                    col,
                    direction: placeDirection,
                } = formattedPlaceCommand(text);
                if (isWithinGrid(row, col)) {
                    setPacMan({
                        row: Number(row),
                        col: Number(col),
                        direction: placeDirection,
                    });
                    break;
                }

                logCommandHistory(
                    `<p>${'Uh Oh cannot move there!'.toUpperCase()}</p>`,
                );
                break;
            }
            case text === 'move': {
                let { col, row } = pacMan;
                const { direction } = pacMan;
                if (direction === 'east') col -= 1;
                else if (direction === 'west') col += 1;
                else if (direction === 'north') row += 1;
                else if (direction === 'south') row -= 1;

                if (isWithinGrid(row, col)) {
                    setPacMan({
                        row: Number(row),
                        col: Number(col),
                        direction,
                    });
                    break;
                }

                logCommandHistory(
                    `<p>${'Uh Oh cannot move there!'.toUpperCase()}</p>`,
                );
                break;
            }
            default: {
                // eslint-disable-next-line no-console
                console.log('Command not found');
            }
        }
    }

    function handleKeyPress(e: KeyboardEvent<HTMLInputElement>) {
        if (e.key === 'Enter') {
            const command = e.currentTarget.value.toLowerCase();
            logCommandHistory(command);
            checkCommand(command);
        }
    }

    return (
        <div className="terminal-container">
            <div className="terminal-result" ref={terminalResultRef}></div>
            <input
                className="terminal-input"
                type="text"
                onKeyDown={handleKeyPress}
                ref={inputRef}
                onInput={(e) => {
                    e.currentTarget.value =
                        `${e.currentTarget.value}`.toUpperCase();
                }}
                placeholder="Try typing MOVE | RIGHT | LEFT | PLACE (0,0,WEST)..."
            />
        </div>
    );
};
