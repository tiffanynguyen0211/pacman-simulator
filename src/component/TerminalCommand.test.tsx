import axe from 'axe-core';
import { mount, ReactWrapper, shallow, ShallowWrapper } from 'enzyme';
import toJson from 'enzyme-to-json';
import { ReactElement } from 'react';

import { mountToDoc } from '../../.jest/jest-setup';
import { PacManPosition } from '../utils/types';
import { TerminalCommand } from './TerminalCommand';

const setup = (
    render: (x: ReactElement) => ReactWrapper | ShallowWrapper,
    props: JSX.IntrinsicAttributes,
) => {
    const defaultProps: {
        pacMan: PacManPosition;
        setPacMan: (newPacmanPosition: PacManPosition) => void;
    } = {
        pacMan: { col: 0, row: 0, direction: 'west' },
        setPacMan: jest.fn().mockImplementation((value) => value),
    };

    const component = render(<TerminalCommand {...defaultProps} {...props} />);

    return {
        actual: component,
    };
};

describe('Terminal Command', () => {
    it('should match snapshot on first render', () => {
        const { actual } = setup(shallow, {});
        expect(toJson(actual)).toMatchSnapshot();
    });

    it('should pass first test case', () => {
        const { actual } = setup(mount, {});
        actual.update();
        const input = actual.find('input');
        const inputEl = input.getDOMNode<HTMLInputElement>();
        const resultLog = actual.find('.terminal-result');

        inputEl.value = 'PLACE 0,0,NORTH';
        input.simulate('keyDown', { key: 'Enter' });
        expect(actual.prop('setPacMan')).toHaveBeenCalledTimes(1);
        let updatedPacman = (actual.prop('setPacMan') as jest.Mock).mock
            .results[0].value;
        actual.setProps({ pacMan: updatedPacman });
        actual.update();

        inputEl.value = 'MOVE';
        input.simulate('keyDown', { key: 'Enter' });
        expect(actual.prop('setPacMan')).toHaveBeenCalledTimes(2);
        updatedPacman = (actual.prop('setPacMan') as jest.Mock).mock.results[1]
            .value;
        actual.setProps({ pacMan: updatedPacman });
        actual.update();

        inputEl.value = 'REPORT';
        input.simulate('keyDown', { key: 'Enter' });

        expect(resultLog.getDOMNode().innerHTML).toMatch(/OUTPUT: 0,1,NORTH/);
    });

    it('should pass second test case', () => {
        const { actual } = setup(mount, {});
        actual.update();
        const input = actual.find('input');
        const inputEl = input.getDOMNode<HTMLInputElement>();
        const resultLog = actual.find('.terminal-result');

        inputEl.value = 'PLACE 0,0,NORTH';
        input.simulate('keyDown', { key: 'Enter' });
        expect(actual.prop('setPacMan')).toHaveBeenCalledTimes(1);
        let updatedPacman = (actual.prop('setPacMan') as jest.Mock).mock
            .results[0].value;
        actual.setProps({ pacMan: updatedPacman });
        actual.update();

        inputEl.value = 'LEFT';
        input.simulate('keyDown', { key: 'Enter' });
        expect(actual.prop('setPacMan')).toHaveBeenCalledTimes(2);
        updatedPacman = (actual.prop('setPacMan') as jest.Mock).mock.results[1]
            .value;
        actual.setProps({ pacMan: updatedPacman });
        actual.update();

        inputEl.value = 'REPORT';
        input.simulate('keyDown', { key: 'Enter' });

        expect(resultLog.getDOMNode().innerHTML).toMatch(/OUTPUT: 0,0,WEST/);
    });

    it('should pass third test case', () => {
        const { actual } = setup(mount, {});
        actual.update();
        const input = actual.find('input');
        const inputEl = input.getDOMNode<HTMLInputElement>();
        const resultLog = actual.find('.terminal-result');

        inputEl.value = 'PLACE 1,2,EAST';
        input.simulate('keyDown', { key: 'Enter' });
        expect(actual.prop('setPacMan')).toHaveBeenCalledTimes(1);
        let updatedPacman = (actual.prop('setPacMan') as jest.Mock).mock
            .results[0].value;
        actual.setProps({ pacMan: updatedPacman });
        actual.update();

        inputEl.value = 'MOVE';
        input.simulate('keyDown', { key: 'Enter' });
        expect(actual.prop('setPacMan')).toHaveBeenCalledTimes(2);
        updatedPacman = (actual.prop('setPacMan') as jest.Mock).mock.results[1]
            .value;
        actual.setProps({ pacMan: updatedPacman });
        actual.update();

        inputEl.value = 'MOVE';
        input.simulate('keyDown', { key: 'Enter' });
        expect(actual.prop('setPacMan')).toHaveBeenCalledTimes(3);
        updatedPacman = (actual.prop('setPacMan') as jest.Mock).mock.results[2]
            .value;
        actual.setProps({ pacMan: updatedPacman });
        actual.update();

        inputEl.value = 'LEFT';
        input.simulate('keyDown', { key: 'Enter' });
        expect(actual.prop('setPacMan')).toHaveBeenCalledTimes(4);
        updatedPacman = (actual.prop('setPacMan') as jest.Mock).mock.results[3]
            .value;
        actual.setProps({ pacMan: updatedPacman });
        actual.update();

        inputEl.value = 'move';
        input.simulate('keyDown', { key: 'Enter' });
        expect(actual.prop('setPacMan')).toHaveBeenCalledTimes(5);
        updatedPacman = (actual.prop('setPacMan') as jest.Mock).mock.results[4]
            .value;
        actual.setProps({ pacMan: updatedPacman });
        actual.update();

        inputEl.value = 'REPORT';
        input.simulate('keyDown', { key: 'Enter' });

        expect(resultLog.getDOMNode().innerHTML).toMatch(/OUTPUT: 0,2,NORTH/);
    });

    it('has no accessibility violations', () => {
        const { actual } = setup(mountToDoc, {});
        const componentNode = actual.getDOMNode();

        axe.run(componentNode, (err, { violations }) => {
            expect(err).toBe(null);
            expect(violations).toHaveLength(0);
        });
    });
});
