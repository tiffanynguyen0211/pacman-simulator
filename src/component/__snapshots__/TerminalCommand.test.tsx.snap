// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Terminal Command should match snapshot on first render 1`] = `
<div
  className="terminal-container"
>
  <div
    className="terminal-result"
  />
  <input
    className="terminal-input"
    onInput={[Function]}
    onKeyDown={[Function]}
    placeholder="Try typing MOVE | RIGHT | LEFT | PLACE (0,0,WEST)..."
    type="text"
  />
</div>
`;
